edit/completers.go

```diff
@@ -111,29 +116,34 @@ func complVariable(n parse.Node, ev *eval.Evaler) (*complSpec, error) {
 		ns = ns[:len(ns)-1]
 	}
 
-	// Collect matching variables.
-	var entries []string
-	iterateVariables(ev, ns, func(varname string) {
-		entries = append(entries, varname)
-	})
-	// Collect namespace prefixes.
-	// TODO Support non-module namespaces.
-	for mod := range ev.Modules {
-		modNsPart := mod + ":"
-		// This is to match namespaces that are "nested" under the current
-		// namespace.
-		if hasProperPrefix(modNsPart, nsPart) {
-			entries = append(entries, modNsPart[len(nsPart):])
+	rawCands := make(chan eval.Value)
+	collectDone := make(chan struct{})
+	go func() {
+		defer func() {
+			close(rawCands)
+			collectDone <- struct{}{}
+		}()
+
+		// Collect matching variables.
+		iterateVariables(ev, ns, func(varname string) {
+			rawCands <- noQuoteCandidate(varname)
+		})
+
+		// Collect namespace prefixes.
+		// TODO Support non-module namespaces.
+		for mod := range ev.Modules {
+			modNsPart := mod + ":"
+			// This is to match namespaces that are "nested" under the current
+			// namespace.
+			if hasProperPrefix(modNsPart, nsPart) {
+				rawCands <- noQuoteCandidate(modNsPart[len(nsPart):])
+			}
 		}
-	}
-	sort.Strings(entries)
+	}()
 
-	var rawCands []rawCandidate
-	for _, entry := range entries {
-		rawCands = append(rawCands, noQuoteCandidate(entry))
-	}
-	cands, err := ev.Editor.(*Editor).filterAndCookCandidates(ev, "variable", nameHead,
+	cands, err := ev.Editor.(*Editor).filterAndCookCandidates(ev, matcher, nameHead,
 		rawCands, parse.Bareword)
+	<-collectDone
```

Comment:
Can you explain why this synchronization is necessary? If it is, consider using `close(collectDone)` above, it is the style used throughout the rest of the program.